smartcontract/service/wasmvm/contract.go

```diff
@@ -0,0 +1,223 @@
+		panic(err)
+	}
+
+	cost := CONTRACT_CREATE_GAS + uint64(uint64(codeLen)/PER_UNIT_CODE_LEN)*UINT_DEPLOY_CODE_LEN_GAS
+	self.checkGas(cost)
+
+	name := make([]byte, nameLen)
+	_, err = proc.ReadAt(name, int64(namePtr))
+	if err != nil {
+		panic(err)
+	}
+
+	version := make([]byte, verLen)
+	_, err = proc.ReadAt(version, int64(verPtr))
+	if err != nil {
+		panic(err)
+	}
+
+	author := make([]byte, authorLen)
+	_, err = proc.ReadAt(author, int64(authorPtr))
+	if err != nil {
+		panic(err)
+	}
+
+	email := make([]byte, emailLen)
+	_, err = proc.ReadAt(email, int64(emailPtr))
+	if err != nil {
+		panic(err)
+	}
+
+	desc := make([]byte, descLen)
+	_, err = proc.ReadAt(desc, int64(descPtr))
+	if err != nil {
+		panic(err)
+	}
+
+	dep, err := payload.CreateDeployCode(code, needStorage, name, version, author, email, desc)
+	if err != nil {
+		panic(err)
+	}
+
+	contractAddr := dep.Address()
+	if self.isContractExist(contractAddr) {
+		panic(errors.NewErr("contract has been deployed"))
+	}
+
+	err = self.Service.CacheDB.PutContract(dep)
+	if err != nil {
+		panic(err)
+	}
+
+	length, err := proc.WriteAt(contractAddr[:], int64(newAddressPtr))
+	return uint32(length)
+
+}
+
+func (self *Runtime) ContractMigrate(proc *exec.Process,
+	codePtr uint32,
+	codeLen uint32,
+	needStorage uint32,
+	namePtr uint32,
+	nameLen uint32,
+	verPtr uint32,
+	verLen uint32,
+	authorPtr uint32,
+	authorLen uint32,
+	emailPtr uint32,
+	emailLen uint32,
+	descPtr uint32,
+	descLen uint32,
+	newAddressPtr uint32) uint32 {
+
+	if uint32(proc.MemAllocated()) < codeLen+nameLen+verLen+authorLen+emailLen+descLen {
+		panic(errors.NewErr("contract migrate len is greater than memory size"))
+	}
+
+	code := make([]byte, codeLen)
+	_, err := proc.ReadAt(code, int64(codePtr))
+	if err != nil {
+		panic(err)
+	}
+
+	cost := CONTRACT_CREATE_GAS + uint64(uint64(codeLen)/PER_UNIT_CODE_LEN)*UINT_DEPLOY_CODE_LEN_GAS
+	self.checkGas(cost)
+
+	name := make([]byte, nameLen)
+	_, err = proc.ReadAt(name, int64(namePtr))
+	if err != nil {
+		panic(err)
+	}
+
+	version := make([]byte, verLen)
+	_, err = proc.ReadAt(version, int64(verPtr))
+	if err != nil {
+		panic(err)
+	}
+
+	author := make([]byte, authorLen)
+	_, err = proc.ReadAt(author, int64(authorPtr))
+	if err != nil {
+		panic(err)
+	}
+
+	email := make([]byte, emailLen)
+	_, err = proc.ReadAt(email, int64(emailPtr))
+	if err != nil {
+		panic(err)
+	}
+
+	desc := make([]byte, descLen)
+	_, err = proc.ReadAt(desc, int64(descPtr))
+	if err != nil {
+		panic(err)
+	}
+
+	dep, err := payload.CreateDeployCode(code, needStorage, name, version, author, email, desc)
+	if err != nil {
+		panic(err)
+	}
+
+	contractAddr := dep.Address()
+	if self.isContractExist(contractAddr) {
+		panic(errors.NewErr("contract has been deployed"))
+	}
+	oldAddress := self.Service.ContextRef.CurrentContext().ContractAddress
+
+	iter := self.Service.CacheDB.NewIterator(oldAddress[:])
+	for has := iter.First(); has; has = iter.Next() {
+		key := iter.Key()
```

Comment:
Does checkGas need?

Reply:
followed by neovm , not charge the storage migration gas